resources/idlharness-shadowrealm.js

```diff
@@ -0,0 +1,78 @@
+// TODO: it would be nice to support `idl_array.add_objects`
+/**
+ * idl_test_shadowrealm is a promise_test wrapper that handles the fetching of the IDL, and
+ * running the code in a `ShadowRealm`, avoiding repetitive boilerplate.
+ *
+ * @see https://github.com/tc39/proposal-shadowrealm
+ * @param {String[]} srcs Spec name(s) for source idl files (fetched from
+ *      /interfaces/{name}.idl).
+ * @param {String[]} deps Spec name(s) for dependency idl files (fetched
+ *      from /interfaces/{name}.idl). Order is important - dependencies from
+ *      each source will only be included if they're already know to be a
+ *      dependency (i.e. have already been seen).
+ */
+function idl_test_shadowrealm(srcs, deps) {
+    const script_urls = [
+        "/resources/testharness.js",
+        "/resources/WebIDLParser.js",
+        "/resources/idlharness.js",
+    ];
+    promise_test(async t => {
+        const realm = new ShadowRealm();
+        realm.evaluate("globalThis.self = globalThis; undefined");
+
+        const ss = await Promise.all(script_urls.map(url => fetch_text(url)));
+        for (const s of ss) {
+            realm.evaluate(s);
+        }
+        const specs = await Promise.all(srcs.concat(deps).map(fetch_spec));
+        const idls = JSON.stringify(specs.map(i => i.idl));
+        const code = `
+            const idls = ${idls};
+            let results;
+            add_completion_callback(function (tests, harness_status, asserts_run) {
+                results = tests;
+            });
+
+            // Without the wrapping test, testharness.js will think it's done after it has run
+            // the first idlharness test.
+            test(() => {
+                const idl_array = new IdlArray();
+                for (let i = 0; i < ${srcs.length}; i++) {
+                    idl_array.add_idls(idls[i]);
+                }
+                for (let i = ${srcs.length}; i < ${srcs.length + deps.length}; i++) {
+                    idl_array.add_dependency_idls(idls[i]);
+                }
+                idl_array.test();
+            }, "inner setup");
+            String(JSON.stringify(results))
+        `;
+
+        // We ran the tests in the ShadowRealm and gathered the results. Now treat them as if
+        // we'd run them directly here, so we can see them.
+        const results = JSON.parse(realm.evaluate(code));
+        for (const {name, status, message} of results) {
+            // TODO: make this an API in testharness.js - needs RFC?
```

Comment:
Yes, this is rather cryptic. Is this how results from a worker context are transferred too? Is the use of `async_test` just to avoid something obscure about the timing when testharness.js considers all results complete?

Reply:
This was basically a suggestion from James. I'm not sure if `test()` worked - in any case, after `promise_setup`, only `promise_test` is supported, so that's what it uses now. The goal is to move this to th.js Soon(TM), probably in early January, but I'd like to land this in its current hack state before that.

Reply:
Oh, delightful. I'm going to go 👀 and this TODO will have to do.